src/app/CommandSender.h

```diff
@@ -36,15 +38,71 @@
 #include <system/SystemPacketBuffer.h>
 
 #include <app/Command.h>
+#include <app/MessageDef/CommandPath.h>
+#include <app/MessageDef/StatusElement.h>
 
 #define COMMON_STATUS_SUCCESS 0
 
 namespace chip {
 namespace app {
 
-class CommandSender : public Command, public Messaging::ExchangeDelegate
+class CommandSender final : public Command, public Messaging::ExchangeDelegate
 {
 public:
+    class Callback
+    {
+    public:
+        virtual ~Callback() = default;
+
+        /**
+         * OnResponse will be called when the response from server arrived.
+         *  - When StatusElement is received and it is Success, aData will be nullptr.
+         *  - When command specific response is received, aData will be a valid TLVReader initialized with a struct of command
+         * specify response.
+         *  - The ProtocolCode in StatusElement MUST be Success if this function is called.
+         *
+         * @param[in] apCommandSender: The command sender object that initialized the command transaction, MUST NOT be freed in this
+         * callback.
+         * @param[in] aPath: The command path field in invoke command response.
+         * @param[in] aData: The command data, will be nullptr if the server returns a StatusElement.
+         */
+        virtual void OnResponse(const CommandSender * apCommandSender, const CommandPath::Type & aPath, TLV::TLVReader * aData) {}
+
+        /**
+         * OnError will be called when error occurred when receiving / handling response from server:
+         * - CHIP_ERROR_TIMEOUT: The CommandSender failed to receive a expected response.
+         * - TLV Related erros: The CommandSender received a malformed response from server.
+         * - CHIP_NO_ERROR: The CommandSender received a legal response from server, but the protocol code indicates a error status
```

Comment:
It's a little counter-intuitive to receive a CHIP_NO_ERROR (which indicates success), but still receive a protocol status code indicating an error.

I'd rather we return a CHIP error like (CHIP_IM_ERROR) that more clearly indicates that there is a more detailed error code that will convey the status.

In the future, we really should have IM be a sub-component in the `SdkPart` within the CHIP error space, allowing us to fold this down to just CHIP_ERROR.